File: tests/cli/setup.test.js
/** setup writes AI settings and installs a missing browser without launching one. */
const { test } = require('node:test');
const assert = require('node:assert/strict');
const { execFile } = require('node:child_process');
const { promisify } = require('node:util');
const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');
const { VIBIUM } = require('../helpers');
const exec = promisify(execFile);

function environment(t, extra = {}) {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'vs-setup-'));
  t.after(() => fs.rmSync(dir, { recursive: true, force: true }));
  const env = {
    ...process.env, HOME: dir, USERPROFILE: dir, VIBIUM_CACHE_DIR: path.join(dir, 'cache'),
    VIBIUM_SESSION: 'setup-cli', VIBIUM_ENGINE: 'chrome', VIBIUM_ENGINE_PATH: '',
    VIBIUM_ENGINE_CHANNEL: '', VIBIUM_ENGINE_VERSION: '', VIBIUM_CONNECT_URL: '',
    ...extra,
  };
  for (const key of ['VIBIUM_AI_PROVIDER', 'VIBIUM_AI_MODEL', 'OPENAI_API_KEY',
    'ANTHROPIC_API_KEY', 'GOOGLE_API_KEY', 'VIBIUM_AI_BASE_URL', 'VIBIUM_AI_REASONING_EFFORT']) {
    if (!(key in extra)) delete env[key];
  }
  return env;
}

async function run(env, args) {
  try { return { ...(await exec(VIBIUM, args, { env, timeout: 15000 })), code: 0 }; }
  catch (error) { if (typeof error.code !== 'number') throw error; return error; }
}

function seedChrome(env) {
  const version = path.join(env.VIBIUM_CACHE_DIR, 'chrome-for-testing', '150.0.0.1');
  const chrome = path.join(version, process.platform === 'darwin'
    ? 'Google Chrome for Testing.app/Contents/MacOS/Google Chrome for Testing'
    : process.platform === 'win32' ? 'chrome.exe' : 'chrome');
  const driver = path.join(version, process.platform === 'win32' ? 'chromedriver.exe' : 'chromedriver');
  env.READY_LAUNCH_MARKER = path.join(env.HOME, 'unexpected-browser-launch');
  for (const executable of [chrome, driver]) {
    fs.mkdirSync(path.dirname(executable), { recursive: true });
    fs.writeFileSync(executable, '#!/bin/sh\nprintf launched > "$READY_LAUNCH_MARKER"\nexit 91\n', { mode: 0o700 });
  }
  return { chrome, driver };
}

test('setup --non-interactive --json with a seeded browser cache returns an ok envelope', async t => {
  const env = environment(t);
  seedChrome(env);
  const result = await run(env, ['setup', '--non-interactive', '--json']);
  assert.equal(result.code, 0, result.stdout + result.stderr);
  const body = JSON.parse(result.stdout);
  assert.equal(body.ok, true);
  assert.deepEqual(body.result.sections.map(s => s.name), ['ai', 'skills', 'browser'],
    'prompt-driven sections must run before the browser download');
  const sections = Object.fromEntries(body.result.sections.map(s => [s.name, s]));
  assert.equal(sections.browser.status, 'done');
  assert.equal(sections.ai.status, 'skipped');
  assert.equal(sections.skills.status, 'skipped');
  assert.equal(fs.existsSync(env.READY_LAUNCH_MARKER), false, 'launched a browser');
});

test('setup without an API key ends on the next step, not a failure', async t => {
  const env = environment(t);
  seedChrome(env);
  const settings = path.join(env.HOME, '.config', 'vibium', 'ai.env');
  fs.mkdirSync(path.dirname(settings), { recursive: true });
  fs.writeFileSync(settings, 'export VIBIUM_AI_PROVIDER=openai\nexport VIBIUM_AI_MODEL=gpt-test\n', { mode: 0o600 });
  const result = await run(env, ['setup', '--non-interactive', '--json']);
  assert.equal(result.code, 0, result.stdout + result.stderr);
  const body = JSON.parse(result.stdout);
  assert.equal(body.ok, true);
  const cred = body.result.ready.checks.find(c => c.name === 'OPENAI_API_KEY');
  assert.equal(cred.status, 'skipped');
  assert.match(cred.message, /add it to .*ai\.env/);
  const provider = body.result.ready.checks.find(c => c.name === 'provider');
  assert.equal(provider.status, 'skipped');
  assert.match(provider.message, /No API key; provider was not contacted/);
  assert.equal(body.result.ready.checks.filter(c => c.status === 'failed').length, 0);
  assert.match(body.result.ready.summary, /Add your API key to .*ai\.env, then run vibium ready ai/);
});

test('setup rejects an unsupported channel instead of ignoring it', async t => {
  const env = environment(t);
  seedChrome(env);
  const result = await run(env, ['setup', '--channel', 'bogus', '--non-interactive', '--json']);
  assert.notEqual(result.code, 0);
  assert.match(result.stdout + result.stderr, /unsupported channel "bogus"/);
});

test('setup does not overwrite an existing ai.env', async t => {
  const env = environment(t);
  seedChrome(env);
  const settings = path.join(env.HOME, '.config', 'vibium', 'ai.env');
  fs.mkdirSync(path.dirname(settings), { recursive: true });
  const original = 'export VIBIUM_AI_PROVIDER=openai\nexport VIBIUM_AI_MODEL=keep-me\n';
  fs.writeFileSync(settings, original, { mode: 0o600 });
  const result = await run(env, ['setup', 'ai', '--non-interactive', '--json']);
  assert.equal(result.code, 0, result.stdout + result.stderr);
  assert.equal(fs.readFileSync(settings, 'utf8'), original);
  assert.equal(fs.existsSync(settings + '.bak'), false);
  const body = JSON.parse(result.stdout);
  assert.equal(body.result.sections.find(s => s.name === 'ai').status, 'skipped');
});

File: clicker/cmd/clicker/setup_test.go
package main

import (
	"bytes"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"

	"github.com/spf13/cobra"
)

func setupTestEnv(t *testing.T) string {
	t.Helper()
	home := t.TempDir()
	t.Setenv("HOME", home)
	t.Setenv("USERPROFILE", home)
	t.Setenv("VIBIUM_CACHE_DIR", filepath.Join(home, "cache"))
	t.Setenv("VIBIUM_CONFIG_DIR", "")
	for _, key := range []string{"VIBIUM_AI_PROVIDER", "VIBIUM_AI_MODEL", "VIBIUM_AI_BASE_URL", "VIBIUM_AI_REASONING_EFFORT", "OPENAI_API_KEY", "ANTHROPIC_API_KEY", "GOOGLE_API_KEY", "VIBIUM_ENGINE", "VIBIUM_ENGINE_CHANNEL", "VIBIUM_ENGINE_PATH", "VIBIUM_SKIP_BROWSER_DOWNLOAD"} {
		t.Setenv(key, "")
	}
	oldJSON, oldEngine, oldChannel := jsonOutput, engineName, engineChannel
	t.Cleanup(func() { jsonOutput, engineName, engineChannel = oldJSON, oldEngine, oldChannel })
	jsonOutput, engineName, engineChannel = true, "chrome", ""
	return home
}

func seedChrome(t *testing.T, home string) {
	t.Helper()
	cache := filepath.Join(home, "cache")
	version := filepath.Join(cache, "chrome-for-testing", "999.0.0.0")
	chrome := filepath.Join(version, "chrome")
	if runtime.GOOS == "darwin" {
		chrome = filepath.Join(version, "Google Chrome for Testing.app", "Contents", "MacOS", "Google Chrome for Testing")
	} else if runtime.GOOS == "windows" {
		chrome = filepath.Join(version, "chrome.exe")
	}
	if err := os.MkdirAll(filepath.Dir(chrome), 0o755); err != nil {
		t.Fatal(err)
	}
	driver := filepath.Join(version, "chromedriver")
	if runtime.GOOS == "windows" {
		driver = filepath.Join(version, "chromedriver.exe")
	}
	for _, p := range []string{chrome, driver} {
		if err := os.WriteFile(p, []byte("fixture"), 0o700); err != nil {
			t.Fatal(err)
		}
	}
}

func quietUI() *setupUI {
	return &setupUI{in: &bytes.Buffer{}, out: ioDiscard(), err: ioDiscard(), interactive: false}
}

func ioDiscard() *bytes.Buffer { return &bytes.Buffer{} }

func TestSetupAINonInteractiveSkipsWhenMissingAndDoesNotOverwrite(t *testing.T) {
	home := setupTestEnv(t)
	cmd := &cobra.Command{}
	sec := setupAI(cmd, quietUI(), false)
	if sec.Status != "skipped" {
		t.Fatalf("missing file: %+v", sec)
	}
	path := filepath.Join(home, ".config", "vibium", "ai.env")
	if _, err := os.Stat(path); err == nil {
		t.Fatal("non-interactive setup wrote ai.env")
	}

	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		t.Fatal(err)
	}
	original := "export VIBIUM_AI_PROVIDER=openai\nexport VIBIUM_AI_MODEL=keep-me\n"
	if err := os.WriteFile(path, []byte(original), 0o600); err != nil {
		t.Fatal(err)
	}
	sec = setupAI(cmd, quietUI(), false)
	if sec.Status != "skipped" {
		t.Fatalf("existing file: %+v", sec)
	}
	got, _ := os.ReadFile(path)
	if string(got) != original {
		t.Fatalf("ai.env was rewritten: %q", got)
	}
	if _, err := os.Stat(path + ".bak"); err == nil {
		t.Fatal("unexpected backup")
	}
}

func TestSetupSkillsOnlyWhenAgentDirExists(t *testing.T) {
	home := setupTestEnv(t)
	cmd := &cobra.Command{}
	sec := setupSkills(cmd, quietUI(), false)
	if sec.Status != "skipped" {
		t.Fatalf("%+v", sec)
	}
	if _, err := os.Stat(filepath.Join(home, ".grok", "skills", "browser", "SKILL.md")); err == nil {
		t.Fatal("installed grok skills without agent dir")
	}

	if err := os.MkdirAll(filepath.Join(home, ".grok"), 0o755); err != nil {
		t.Fatal(err)
	}
	sec = setupSkills(cmd, quietUI(), false)
	if sec.Status != "done" || sec.Agent != "grok" {
		t.Fatalf("%+v", sec)
	}
	if _, err := os.Stat(filepath.Join(home, ".grok", "skills", "browser", "SKILL.md")); err != nil {
		t.Fatal(err)
	}
	if _, err := os.Stat(filepath.Join(home, ".grok", "skills", "check", "SKILL.md")); err != nil {
		t.Fatal(err)
	}
}

func TestSetupQuickSkipsInstalledBrowser(t *testing.T) {
	home := setupTestEnv(t)
	seedChrome(t, home)
	sec := setupBrowser(&cobra.Command{}, quietUI(), true)
	if sec.Status != "skipped" {
		t.Fatalf("%+v", sec)
	}
}

func TestSetupBrowserReportsPresent(t *testing.T) {
	home := setupTestEnv(t)
	seedChrome(t, home)
	sec := setupBrowser(&cobra.Command{}, quietUI(), false)
	if sec.Status != "done" || !strings.Contains(sec.Message, "already installed") {
		t.Fatalf("%+v", sec)
	}
}

func TestWriteAIEnvModeAndBackup(t *testing.T) {
	home := setupTestEnv(t)
	path := filepath.Join(home, ".config", "vibium", "ai.env")
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(path, []byte("export OPENAI_API_KEY='old'\n"), 0o600); err != nil {
		t.Fatal(err)
	}
	if err := writeAIEnv(path, map[string]string{"VIBIUM_AI_PROVIDER": "openai", "VIBIUM_AI_MODEL": "m", "OPENAI_API_KEY": "new"}); err != nil {
		t.Fatal(err)
	}
	info, err := os.Stat(path)
	if err != nil {
		t.Fatal(err)
	}
	if runtime.GOOS != "windows" && info.Mode().Perm() != 0o600 {
		t.Fatalf("mode %o", info.Mode().Perm())
	}
	bak, err := os.ReadFile(path + ".bak")
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(string(bak), "old") {
		t.Fatal("backup lost previous contents")
	}
	body, _ := os.ReadFile(path)
	if strings.Contains(string(body), "old") || !strings.Contains(string(body), "export VIBIUM_AI_PROVIDER='openai'") {
		t.Fatalf("written file: %s", body)
	}
}

func TestSetupTryHintOnlyWhenFullyReady(t *testing.T) {
	setupTestEnv(t)
	if setupTryHint(nil) != "" {
		t.Fatal("hint without readiness")
	}
	if setupTryHint(&setupResult{Ready: true}) != "" {
		t.Fatal("hint without AI configuration")
	}
	t.Setenv("VIBIUM_AI_PROVIDER", "openai")
	t.Setenv("VIBIUM_AI_MODEL", "gpt-test")
	t.Setenv("OPENAI_API_KEY", "k")
	if setupTryHint(&setupResult{Ready: false}) != "" {
		t.Fatal("hint on a not-ready run")
	}
	if hint := setupTryHint(&setupResult{Ready: true}); !strings.Contains(hint, "vibium run") {
		t.Fatalf("hint: %q", hint)
	}
}

func TestParseProviderChoice(t *testing.T) {
	// Digits map onto setupProviders; the accepted range must follow the
	// list so adding a provider never silently truncates the menu.
	for i, want := range setupProviders {
		got, err := parseProviderChoice(string(rune('1'+i)), "openai")
		if err != nil || got != want {
			t.Fatalf("choice %d: got %q %v", i+1, got, err)
		}
	}
	got, err := parseProviderChoice("2", "openai")
	if err != nil || got != "xai" {
		t.Fatalf("got %q %v", got, err)
	}
	got, err = parseProviderChoice("local", "openai")
	if err != nil || got != "local" {
		t.Fatalf("got %q %v", got, err)
	}
	if _, err := parseProviderChoice(string(rune('1'+len(setupProviders))), "openai"); err == nil {
		t.Fatal("accepted digit past the menu")
	}
	if _, err := parseProviderChoice("nope", "openai"); err == nil {
		t.Fatal("accepted unknown provider")
	}
}
